### PR TITLE
Blog post: Hardware Accelerated Crypto...

### PR DESCRIPTION
**Pull request by mythi**
This adds a story of Hardware Accelerated SSL/TLS Termination in Ingress Controllers using Kubernetes Device Plugins and RuntimeClass.

Pre-approved by @castrojo.

Q: How to add the images?

/assign kbarnard10

### CODE REVIEW

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  * Haproxy-ingress deployment `.yaml` to
+     * Request `qat.intel.com: n` resources
+     * Request `runtimeClassName: kata-containers` (name value depends on cluster config)
+  * (QAT device config file for each requested device resource with OpenSSL engine configured available in the container)
+
+Once the building blocks are available, the hardware accelerated SSL/TLS can be tested by following the [TLS termination
+example](https://github.com/jcmoraisjr/haproxy-ingress/tree/master/examples/tls-termination) steps. In order to verify the hardware is used, you can check `/sys/kernel/debug/*/fw_counters` files on host as they
+get updated by the Intel&reg; QAT firmware.
+
+*Haproxy-ingress and HAproxy are used because HAproxy can be directly configured to use the OpenSSL engine using
```
Comment (sftim) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: This `*` looks a little like the start of Markdown emphasis.

Something like:
```markdown
*<span/>Haproxy-ingress and HAproxy are used…
```

might help if emphsis isn't wanted.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+accelerated processing in SSL operations as discussed, and can provide other crypto
+services like key/random number generation. Clouds can make the hardware easily available
+using the dynamic ENGINE and several loadable module implementations exist, for
+example, [CloudHSM](https://docs.aws.amazon.com/cloudhsm/latest/userguide/openssl-library.html), [IBMCA](https://github.com/opencryptoki/openssl-ibmca), or [QAT Engine](https://github.com/intel/QAT_Engine/).
+
+For Cloud deployments, the ideal scenario is for these modules to be shipped as part of
+the container workload. The workload would get scheduled on a node that provides the
+underlying hardware that the module needs to access. On the other hand, the workloads
+should run the same way and without code modifications regardless of the crypto acceleration
+hardware being available or not. The OpenSSL dynamic engine enables this. Figure 1  below
```
Comment (sftim) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]:
```suggestion
hardware being available or not. The OpenSSL dynamic engine enables this. Figure 1 below
```

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md (hunk 1 above)
Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: @sftim I saw this too after I submitted the PR. The preview shows this just fine.

Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: hmm ```<span/``` does not seem to have any effect.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
+**Authors:** Mikko Ylinen (Intel)
+
```
Comment (kbhawkey) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: Hello @mythi.
- Does your file have front matter? Such as:
```yaml
---
layout: blog
title: 'Pod Priority and Preemption in Kubernetes'
date: 2019-04-16
---
```
- I was unable to see the preview of this file.
- I think `content\en\blog\_posts\future-of-cloud-providers.md` contains an image. May help for your changes here.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+isolation layer. In addition to workload isolation, the Kata Containers VM has the added
+side benefit that the VFIO devices, as `Allocate`’d by the device plugin, can be passed
+through to the container as hardware isolated devices. The only requirement is that the
+Kata Containers kernel has driver for the exposed device enabled.
+
+That’s all it really takes to enable hardware accelerated crypto for container workloads. To summarize:
+
+  1. Cluster needs a device plugin running on the node that provides the hardware,
+  2. Device plugin exposes the hardware to user space using  the VFIO driver,
+  3. Pod requests the device resources and Kata Containers as the RuntimeClass in the PodSpec,
```
Comment (kbhawkey) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: Would omit the `commas` after each sentence and could use an unordered list.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     * Request `qat.intel.com: n` resources
+     * Request `runtimeClassName: kata-containers` (name value depends on cluster config)
+  * (QAT device config file for each requested device resource with OpenSSL engine configured available in the container)
+
+Once the building blocks are available, the hardware accelerated SSL/TLS can be tested by following the [TLS termination
+example](https://github.com/jcmoraisjr/haproxy-ingress/tree/master/examples/tls-termination) steps. In order to verify the hardware is used, you can check `/sys/kernel/debug/*/fw_counters` files on host as they
+get updated by the Intel&reg; QAT firmware.
+
+*<span/>Haproxy-ingress and HAproxy are used because HAproxy can be directly configured to use the OpenSSL engine using
+`ssl-engine <name> [algo ALGOs]` configuration flag without modifications to the global openssl configuration file.
```
Comment (kbhawkey) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: Could you clean up the `*<span/>`? Perhaps this was copied from somewhere else.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  * Kubernetes 1.14 (`RuntimeClass` and `DevicePlugin` feature gates enabled (both are `true` in 1.14)
+  * RuntimeClass ready runtime and Kata Containers configured
+
+### Host configuration:
+  * Intel&reg; QAT driver release with the kernel drivers installed for both host kernel and Kata Containers kernel (or on a rootfs as loadable modules)
+  * [QAT device plugin](https://github.com/intel/intel-device-plugins-for-kubernetes/tree/master/cmd/qat_plugin) DaemonSet deployed
+
+### Ingress controller configuration and deployment:
+  * [HAproxy-ingress](https://github.com/jcmoraisjr/haproxy-ingress)* ingress controller in a modified container that has
+     * the QAT HW HAL user space library (part of Intel&reg; QAT SW release) and
```
Comment (kbhawkey) [inline content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: Do you need the asterisk after `https://github.com/jcmoraisjr/haproxy-ingress)*`

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md (hunk 4 above)
Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: dropped the commas

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md (hunk 6 above)
Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: The asterisk is a footnote. I've now replaced it with ```[^1]```

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md (hunk 3 above)
Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: @kbhawkey thanks for the review. I've added the front matter. Let's see if the preview works after that.

Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: I don't yet have the date for publication agreed. I've picked 04/23 and made changes (including the images) based on that. The preview should work now.

File: content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md (hunk 6 above)
Comment (mythi) [reply content/en/blog/_posts/Hardware-Accelerated-SSLTLS-Termination-in-Ingress-Controllers-using-Kubernetes-Device-Plugins-and-RuntimeClass.md]: ```[^1]``` did not render like I wanted so I've just dropped it completely.

**Pull request merged**